docs/proposals/application-name-identifier.md

```diff
@@ -0,0 +1,397 @@
+---
+title: Change the way application resources are identified
+authors:
+  - "@jannfis"
+sponsors:
+  - TBD
+reviewers:
+  - TBD
+approvers:
+  - TBD
+
+creation-date: 2021-06-07
+last-updated: 2021-06-07
+---
+
+# Change the way application resources are identified
+
+This is a proposal to introduce using a hash value as application identifier
+in the application instance label. This will allow application names longer
+than 63 characters. As an additional goal, we propose to introduce a GUID as
+installation ID that will allow multiple Argo CD instances manage resources
+on the same cluster, without the need for manual reconfiguration.
+
+## Open Questions [optional]
+
+The major open questions are:
+
+* Which hashing algorithm to use? The authors think that `SHA-1` and `FNV-1a`
+  are the best options on the table, with `FNV-1a` being the more performant
+  one and `SHA-1` being the one most resilient against collisions with a
+  compromise for speed. There is a pretty good comparison of non cryptographic
+  hash algorithms on Stack Exchange that suggests that `FNV-1a` might be more
+  than sufficient for our use-cases:
+  https://softwareengineering.stackexchange.com/a/145633
+
```

Comment:
Re: hash choice - 

If we're going to go with SHA-based algorithm, I recommend SHA-256, as SHA-1 [is considered broken](https://en.wikipedia.org/wiki/SHA-1). It doesn't matter for this use case (we're not using the cryptographic properties of a hash, as you note later), but we will constantly get people telling us it's broken and asking us when we will move to another algorithm :smile: .

Reply:
You have a good argument there. I favored the SHA-1 variant for producing the _shorter_ hash values, but I think you are right in that we may want to consider usage of SHA-256 instead. Although we do not need to be secure in cryptographic terms here, probably SHA-1 will become utterly deprecated soon enough (maybe up to the point of being removed from any standard libraries?), and even Git will move away from it. So yeah, it leaves a "bad aftertaste" kind of thing :)

Thanks for bringing it up, I will change it in the proposal accordingly.

Reply:
Looks like we've got two similar discussion threads in this PR. Copying comments from another thread here.

As I understand there are trying to solve three issues:

* allow app with long names
* support two argocd managing apps in the same cluster
* avoid argocd confusion that happened when something else adds `app.kubernetes.io/instance` label

Also, I think additional requirements are: try to keep it human-friendly and try avoid breaking changes. 

To solve I want to propose:

* use annotations instead of labels for tracking to solve long names issue
* use three annotations (e.g. `argo-cd.argoproj.io/application`, `argo-cd.argoproj.io/resourceKey`, ``argo-cd.argoproj.io/installation` ) that includes app name, resource key and argocd installation id 
* allow the user to switch to annotation-based tracking when necessary using new `app.spec.tracking` field. (this [comment](https://github.com/argoproj/argo-cd/pull/6425#discussion_r647726256) tries to justify why new way of tracking should be optional)


@jannfis , @jgwest what do you think?

Reply:
I think we should not provide two different methods for tracking applications, as it will complicate things greatly and probably will cause some confusion among users.

What speaks against using the label for tracking and the annotations for human-readable identifiers together? So the annotations will only serve informational purposes, and the label is used to select and identify resources.

To solve

> avoid argocd confusion that happened when something else adds app.kubernetes.io/instance label

I would rather propose to change the default `app.kubernetes.io/instance` label to something more sensible, Argo CD specific. E.g. `argocd.argoproj.io/instance` - I think the Argo CD Helm chart already sets this by default.

IMHO, I think we shouldn't sacrifice the label in order to enable longer-than-63-chars application names.

Reply:
I'm trying to avoid painful migration for end-users: if Argo CD changes the tracking method then all managed resources  suddenly became out-of-sync. 

However, I agree that two ways if tracking is not great. In order to make it easier to migrate, we should support using "legacy" way of tracking for 1~2 releases.